Tests/ApolloTests/CacheKeyResolutionTests.swift

```diff
@@ -209,4 +209,55 @@ class CacheKeyResolutionTests: XCTestCase {
     expect(actual).to(equal("GreekLetters:δ"))
   }
 
+  func test__schemaConfiguration__givenSingleKeyField_shouldReturnKeyFieldValue() {
+    let Delta = Object(typename: "Dog", implementedInterfaces: [], keyFields: ["id"])
+
+    MockSchemaMetadata.stub_objectTypeForTypeName({ _ in Delta })
+
+    let object: JSONObject = [
+      "__typename": "Dog",
+      "id": "10",
+      "name": "Beagle"
+    ]
+
+    let objectDict = NetworkResponseExecutionSource().opaqueObjectDataWrapper(for: object)
+    let actual = MockSchemaMetadata.cacheKey(for: objectDict)
+
+    expect(actual).to(equal("Dog:10"))
+  }
+  
+  func test__schemaConfiguration__givenMultipleKeyFields_shouldReturnKeyFieldValues() {
+    let Delta = Object(typename: "Dog", implementedInterfaces: [], keyFields: ["id", "name"])
+
+    MockSchemaMetadata.stub_objectTypeForTypeName({ _ in Delta })
+
+    let object: JSONObject = [
+      "__typename": "Dog",
+      "id": "10",
+      "name": "Beagle",
+      "height": 20,
+    ]
+
+    let objectDict = NetworkResponseExecutionSource().opaqueObjectDataWrapper(for: object)
+    let actual = MockSchemaMetadata.cacheKey(for: objectDict)
+
+    expect(actual).to(equal("Dog:10/Beagle"))
```

Comment:
I believe the current version of Apollo Kotlin doesn't have any separator and just concatenates the fields directly.

https://www.apollographql.com/docs/kotlin/caching/declarative-ids#typepolicy

Reply:
@x-sheep that's correct. The logic is [there](https://github.com/apollographql/apollo-kotlin/blob/7668a302eb5c62ca1d3e423ccfefa1142acee1f4/libraries/apollo-normalized-cache-api/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/api/CacheKey.kt#L19). An object key for type "User" with `keyFields: "firstName lastName"` would  be `User:MartinBonnin`

Reply:
Should I change the separator to `+`? I don't have a preference for which character to use, but using no separator doesn't seem like a good idea.

Reply:
> using no separator doesn't seem like a good idea.

Can you ellaborate on that? 

My view of cache ids is they are "write-only". Given an object there is a function to compute the id. But the other way (reconstructing an object from an id) is a lot less clear, I wouldn't necessarily encourage users to do it. Which is what we're implicitly doing if we add a separator. But maybe there's a use case out there?

Another argument is that separators end up using space in the cache. There are a lot of cache references in a cache so any byte we save matters.

Reply:
I was thinking in case of overlap: `Rectangle:39820` could mean `width: 39, height: 820` or `width: 398, height: 20`

There are multiple documentation pages for Apollo iOS that specify how the cache keys are serialized to string, and sometimes the user can reconstruct them directly.

https://www.apollographql.com/docs/ios/caching/cache-transactions#reading-cache-data

Reply:
> I was thinking in case of overlap: Rectangle:39820 could mean width: 39, height: 820 or width: 398, height: 20

Excellent point, we'd have cache cloberring in that case 👍 . So we need a separator indeed 👍 . 

If we want to be 100% safe, it needs to be something that can never appear in a GraphQL input 🤔 . Which is hard (impossible?) considering [string arguments can be any unicode value](https://spec.graphql.org/draft/#SourceCharacter). Or we'd need to encode them as a json array...

Reply:
To avoid cache clobbering entirely, it could be possible to do a few string replacements on each key field (instead of serializing the entire thing to JSON):

* `\` to `\\`
* `+` to `\+`

And then use `+` as a separator, which can no longer appear without a leading backslash in user data. What do you think?

Reply:
> And then use + as a separator, which can no longer appear without a leading backslash in user data. 

Yup, looks like this would work 👍

Reply:
What about a comma (`,`) as a separator? This is the universal separator in a lot of formats and if we do the escaping trick, we can distinguish it from  a user comma. Or are we concerned that `,` might happen too much to be readable?

Reply:
In my opinion, it'd be preferable if the exact calculation of a cache key could be hidden entirely from users, which would mean deprecating the `ReadTransaction.readObject` functions that take a string as a key, and replacing it with taking a CacheReference struct (with different overloads for user-defined and keyFields-defined values).

If the serialization detail was hidden from users, the exact choice of character for `+` (and even `\`) could be something that has a very low chance of appearing in user data (even codepoints below 0x20 could be considered?)

Reply:
Correct, this is the direction we need to move in for `readObject`; we created [issue #3480](https://github.com/apollographql/apollo-ios/issues/3480) a couple weeks ago to track that. Good timing!

Reply:
I've now set `+` as a separator, along with escaping other instances of `+` with a backslash.

Reply:
Thanks for the great discussion here all. I'm happy with the decision to use `+`. @martinbonnin, do you think you will want to update the Apollo Kotlin to do the same? I think it's an improvement, but also has potential to break caching behavior for current users.

Reply:
> do you think you will want to update the Apollo Kotlin to do the same? I think it's an improvement, but also has potential to break caching behavior for current users.

Most likely won't happen in 4.x as it's hugely breaking. Could potentially be something for [`apollo-normalized-cache-incubating`](https://github.com/apollographql/apollo-kotlin-normalized-cache-incubating). Ping @BoD.

Reply:
👍  I've created https://github.com/apollographql/apollo-kotlin-normalized-cache-incubating/issues/80